### PR TITLE
[ceph-fuse] Fh ref might leak at umounting

### PR DESCRIPTION
**Pull request by david-z**
Fixes: #14800

This PR has been tested on my local cluster with hammer and master branch.

Signed-off-by: Zhi Zhang <zhangz.david@outlook.com>

### CODE REVIEW

File: src/client/Client.h [hunk 1]
```diff
@@ -421,6 +421,7 @@ class Client : public Dispatcher, public md_config_obs_t {
   // file handles, etc.
   interval_set<int> free_fd_set;  // unused fds
   ceph::unordered_map<int, Fh*> fd_map;
+  set<Fh*> ll_unclosed_fh;
```
Comment (ukernel) [inline src/client/Client.h]: I think name it as ll_open_fh_set is better.

**Comment by ukernel** [comment]
I think this patch is not enough for handling the force umount case. Each inode in kernel hold one or more ll_ref. In the force umount case, kernel does not call ll_forget() for release these ll_refs

File: src/client/Client.h (hunk 1 above)
Comment (david-z) [inline src/client/Client.h]: ok, done.

**Comment by david-z** [comment]
@ukernel 

Yes, you're right. ll_ref can't be released under some cases in the force umount. But inode has be freed once inode's ref becomes zero, so ll_ref has no use and no harm even if it is not zero in the force umount. What do you think?

**Comment by david-z** [comment]
@dachary 

Last build is failed due to one rbd's unit test (https://jenkins.ceph.com/job/ceph-pull-requests/1704/console). This PR has nothing to do with rbd and also I have run the same unit test which is passed on my local build env. Could you pls take a look? Thanks.

**Comment by ukernel** [comment]
<pre>
void Client::_ll_get(Inode *in)
{
  if (in->ll_ref == 0) {
    in->get();
    if (in->is_dir() && !in->dn_set.empty()) {
      assert(in->dn_set.size() == 1); // dirs can't be hard-linked
      in->get_first_parent()->get(); // pin dentry
    }
  }
  in->ll_get();
  ldout(cct, 20) << "_ll_get " << in < " " << in->ino << " -> " << in->ll_ref << dendl;
}
</pre

the first ll_ref holds the inode reference.

**Comment by david-z** [comment]
Thanks for mentioning this piece of code. 

By further looking at Client::unmount() function, after releasing all the Fhs, it will call _ll_drop_pins() to release all the inodes' ll_ref in the inode_map. I think we only need to ensure the inodes in the inode_map won't  have any leaked reference, then ceph-fuse process won't be hung. 

So does _ll_drop_pins() resolve the ll_ref's concern? Pls let me know if I am wrong. Thanks.

**Comment by ukernel** [comment]
you are right.

Except ll_unclose_fh looks weird , this patch looks good

**Comment by david-z** [comment]
I already changed the name according to your suggestion. Pls take a look.

BTW, build keeps failling due to one rbd unit test..., which passed on my build env.

**Comment by ukernel** [comment]
http://pulpito.ceph.com/zyan-2016-03-02_06:21:11-fs-wip-zyan-testing-testing-basic-mira/

tests passed

**Comment by gregsfortytwo** [comment]
@david-z is there a functional change with this code, or is it just cleaning up memory prior to shutdown?

**Comment by david-z** [comment]
@gregsfortytwo  no functional change here. It is just using a set for ceph-fuse to record opened Fhs and release them before shutdown.

**Comment by gregsfortytwo** [comment]
http://qa-proxy.ceph.com/teuthology/gregf-2016-03-07_23:12:27-fs-greg-fs-testing-3-7-safe---basic-mira/
5 failures; largely infrastructure; all known and consistent at time of testing.

**Pull request merged**